astropy/units/quantity.py

```diff
@@ -1452,29 +1467,50 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
         }
         formats["latex_inline"] = formats["latex"]
 
-        if format not in formats:
-            raise ValueError(f"Unknown format '{format}'")
-        elif format is None:
-            if precision is None:
+        format_spec = None
+        if formatter is not None and not callable(formatter):
+            format_spec = formatter
+            formatter = {"all": lambda x: builtins.format(x, format_spec)}
+
+        if format is None:
+            # format_spec overwrites precision
+            if formatter is None and precision is None:
                 # Use default formatting settings
                 return f"{self.value}{self._unitstr:s}"
-            else:
-                # np.array2string properly formats arrays as well as scalars
-                return (
-                    np.array2string(self.value, precision=precision, floatmode="fixed")
-                    + self._unitstr
+
+            if callable(formatter):
+                formatter = {"all": formatter}
+
+            return (
+                np.array2string(
+                    self.value,
+                    precision=precision,
+                    floatmode="fixed",
+                    formatter=formatter,
                 )
+                + self._unitstr
+            )
+
+        if format not in formats:
+            raise ValueError(f"Unknown format '{format}'")
 
         # else, for the moment we assume format="latex" or "latex_inline".
 
         # Set the precision if set, otherwise use numpy default
         pops = np.get_printoptions()
-        format_spec = f".{precision if precision is not None else pops['precision']}g"
+        if format_spec is None:
+            format_spec = (
+                f".{precision if precision is not None else pops['precision']}g"
+            )
 
         def float_formatter(value):
+            if callable(formatter):
```

Comment:
This is weird. Instead, I'd write:
```
if formatter is None or isinstance(formatter, str):
    if self.dtype.kind == 'c':
        def formatter(value):
             stuff for complex
    else:
        def formatter(value):
            stuff for real
```
Indeed, it looks like this would allow substantial simplification....